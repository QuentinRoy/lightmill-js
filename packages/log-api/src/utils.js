function updateModelPropertiesInPlace(model, callback) {
  for (const [key, prop] of model.properties) {
    const updatedProp = callback(key, prop);
    model.properties.set(key, updatedProp);
  }
}

export function withPartial(_context, target, props) {
  let notFoundProps = new Set(props ?? []);
  updateModelPropertiesInPlace(target, (_key, prop) => {
    if (props == null || props.includes(prop.name)) {
      notFoundProps.delete(prop.name);
      return { ...prop, optional: true };
    }
    return prop;
  });
  if (notFoundProps.size > 0) {
    throw new Error(
      `Partial decorator: properties ${Array.from(notFoundProps).join(', ')} not found`,
    );
  }
}

export const $decorators = { Utils: { withPartial } };
