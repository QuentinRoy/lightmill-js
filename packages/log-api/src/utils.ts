import { getErrorSchema, getOneErrorDocumentSchema } from './jsonapi.ts';
import { z } from './zod-openapi.ts';

export function mapKeys<
  T extends object,
  M extends (k: Extract<keyof T, string>) => PropertyKey,
>(obj: T, fn: M) {
  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => {
      // @ts-expect-error: we assume T has no extra properties so key is
      // a valid key of T.
      return [fn(key), value];
    }),
  ) as { [K in keyof T as ReturnType<M>]: T[K] };
}

export function prefixKeys<T extends Record<string, unknown>, P extends string>(
  obj: T,
  prefix: P,
) {
  return mapKeys(obj, (k) => {
    return `${prefix}${k as keyof T & (string | number)}` as const;
  });
}

// Common types
export const StringOrArrayOfStrings = z
  .union([z.string(), z.array(z.string())])
  .openapi('StringOrArrayOfStrings');

export const ForbiddenErrorResponse = getOneErrorDocumentSchema(
  getErrorSchema({
    code: 'FORBIDDEN',
    statusText: 'Forbidden',
    statusCode: 403,
  }),
).openapi('ForbiddenErrorResponse');
