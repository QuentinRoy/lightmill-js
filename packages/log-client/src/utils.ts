import { JsonValue } from 'type-fest';

type JsonFetchUrl = string | URL;
type JsonFetchOptions<AllowBody extends boolean = true> = Omit<
  NonNullable<Parameters<typeof fetch>[1]>,
  'body' | 'method' | 'headers'
> & {
  body?: AllowBody extends true ? JsonValue : never;
  headers?: Record<string, string>;
};

function makeJsonFetchMethod(
  method: string,
  opt: { noBody: true },
): (url: JsonFetchUrl, options: JsonFetchOptions<false>) => Promise<unknown>;
function makeJsonFetchMethod(
  method: string,
  opt?: { noBody: false },
): (url: JsonFetchUrl, options: JsonFetchOptions) => Promise<unknown>;
function makeJsonFetchMethod(method: string, { noBody = false } = {}) {
  return async function fetchMethod(
    url: JsonFetchUrl,
    options: JsonFetchOptions,
  ) {
    let body = noBody ? undefined : JSON.stringify(options.body);
    let headers: JsonFetchOptions['headers'] = {
      ...options.headers,
      Accept: 'application/json',
    };
    if (body != null) {
      headers['Content-Type'] = 'application/json';
    }
    let response = await fetch(url, { ...options, method, body, headers });
    let json = await response.json();
    if (response.ok) {
      return json as unknown;
    } else {
      throw new Error(json.message ?? 'Unknown error');
    }
  };
}

export const post = makeJsonFetchMethod('POST');
export const put = makeJsonFetchMethod('PUT');
export const get = makeJsonFetchMethod('GET', { noBody: true });
export const del = makeJsonFetchMethod('DELETE');
export const patch = makeJsonFetchMethod('PATCH');
