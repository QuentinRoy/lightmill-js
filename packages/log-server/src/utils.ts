import { mapKeys } from 'remeda';
import { SnakeCase } from 'type-fest';
import { snakeCase } from 'change-case';
import { z } from 'zod';

export function arrayify<T>(value: T | T[], isEmptyWithUndefined?: false): T[];
export function arrayify<T>(
  value: T | T[] | undefined,
  isEmptyWithUndefined: true,
): T[];
export function arrayify<T>(value: T | T[], isEmptyWithUndefined = false): T[] {
  if (value === undefined && isEmptyWithUndefined) return [];
  return Array.isArray(value) ? value : [value];
}

export const JsonPrimitive = z.union([
  z.string(),
  z.number(),
  z.boolean(),
  z.null(),
]);
export type JsonPrimitive = z.infer<typeof JsonPrimitive>;
export type JsonValue =
  | JsonPrimitive
  | ({ [Key in string]: JsonValue } & {
      [Key in string]?: JsonValue | undefined;
    })
  | JsonValue[]
  | readonly JsonValue[];
export const JsonValue: z.ZodType<JsonValue> = z.lazy(() =>
  z.union([JsonPrimitive, z.array(JsonValue), z.record(JsonValue)]),
);
export const JsonObject = z.record(JsonValue);
export type JsonObject = z.infer<typeof JsonObject>;
export const JsonArray = z.array(JsonValue);
export type JsonArray = z.infer<typeof JsonArray>;

export function toSnakeCase<
  R extends Record<string | number | symbol, unknown>,
>(input: R) {
  return mapKeys(input, (key) =>
    typeof key === 'string' ? snakeCase(key) : key,
  ) as unknown as {
    [K in keyof R as SnakeCase<K>]: R[K];
  };
}
