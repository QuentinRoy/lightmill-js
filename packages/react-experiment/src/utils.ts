import { UnionToIntersection } from 'type-fest';

export type IsUnion<T> = [T] extends [UnionToIntersection<T>] ? false : true;
export type NotUnion<T> = true extends IsUnion<T> ? never : T;

export function promisy<T>(
  t: T | (() => T) | (() => Promise<T>),
): () => Promise<T> {
  if (t instanceof Function) {
    return () => Promise.resolve(t());
  }
  return () => Promise.resolve(t);
}
