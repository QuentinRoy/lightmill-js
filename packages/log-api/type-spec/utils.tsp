import "@typespec/http";
import "./json-api.tsp";
import "./utils.js";

using TypeSpec.Http;
using JsonApi;

namespace Utils;

extern dec withPartial(target: Reflection.Model, props?: valueof Array<string>);

model InvalidQueryParameterDoc
  is ErrorDocument<Error<"INVALID_QUERY_PARAMETER", HttpStatus.`400`> & {
    source: {
      parameter: string;
    };
  }>;

/**
 * Error document sent when an operation is rejected due to insufficient
 * permissions.
 */
model ForbiddenErrorDocTemplate<Code extends string = "FORBIDDEN">
  is ErrorDocument<Error<Code, HttpStatus.`403`>>;
model ForbiddenErrorDoc is ForbiddenErrorDocTemplate<"FORBIDDEN">;
model ForbiddenErrorResponse extends ForbiddenResponse {
  @body doc: ForbiddenErrorDoc;
  @header contentType: JsonApi.APIContentType;
}

model InvalidCredentialsError is Error<"INVALID_CREDENTIALS", HttpStatus.`403`>;
model InvalidCredentialsErrorResponse extends ForbiddenResponse {
  @body doc: ErrorDocument<InvalidCredentialsError>;
  @header contentType: JsonApi.APIContentType;
}

model NotFoundError<Code extends string = "NOT_FOUND">
  is Error<Code, HttpStatus.`404`>;
/**
 * Error document sent when a resource is not found.
 */
model NotFoundErrorDocTemplate<Code extends string = "NOT_FOUND">
  is ErrorDocument<NotFoundError<Code>>;
model NotFoundErrorDoc is NotFoundErrorDocTemplate<"NOT_FOUND">;
model NotFoundErrorResponse extends NotFoundResponse {
  @body doc: NotFoundErrorDoc;
  @header contentType: JsonApi.APIContentType;
}

/**
 * Error document sent when a request is made with an invalid session.
 */
model SessionRequiredError is Error<"SESSION_REQUIRED", HttpStatus.`403`>;
model SessionRequiredErrorResponse extends ForbiddenResponse {
  @body doc: ErrorDocument<SessionRequiredError>;
  @header contentType: JsonApi.APIContentType;
}

model UnsupportedMediaTypeError
  is Error<"UNSUPPORTED_MEDIA_TYPE", HttpStatus.`415`>;
/**
 * Error document sent when an unsupported media type is requested.
 */
model UnsupportedMediaTypeErrorDoc
  is ErrorDocument<Error<"UNSUPPORTED_MEDIA_TYPE", HttpStatus.`415`>>;
model UnsupportedMediaTypeErrorResponse extends Response<415> {
  @body doc: ErrorDocument<UnsupportedMediaTypeError>;
  @header contentType: JsonApi.APIContentType;
}

alias CookieSessionName = "lightmill-session-id";

// I don't want this to be in main because I don't want this model to be
// included in the generated OpenAPI schemas.
@doc("Cookie-based session")
model CookieSessionAuth {
  @doc("Http authentication")
  type: AuthType.apiKey;

  @doc("location of the API key")
  in: ApiKeyLocation.cookie;

  @doc("name of the API key")
  name: CookieSessionName;
}

@withPartial
model Partial<T> {
  ...T;
}
