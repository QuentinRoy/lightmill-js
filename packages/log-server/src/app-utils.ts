import type { paths } from '@lightmill/log-api';
import type { SessionData } from 'express-session';
import { groupBy, intersection, map, pipe, uniqueBy } from 'remeda';
import type { Simplify, WritableDeep } from 'type-fest';
import {
  type ApiPath,
  type HttpMethod,
  httpStatuses,
  type HttpStatusMap,
} from './api-utils.js';
import type { RunId, Store } from './store.js';
import type {
  Handler,
  HandlerParameters,
  HandlerResult,
  RequestContent,
  ServerDescription,
} from './typed-server.js';
import { arrayify } from './utils.js';

declare module 'express-session' {
  interface SessionData {
    data: { role: 'participant' | 'host'; runs: RunId[] };
  }
}

// This needs to be a type (not an interface) so we can use it with
// typed-server... I don't know why, but it's not worth investigating.
export type ServerApi = Simplify<paths>;

export type ServerHandler<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = Handler<ServerApi, Path, Method>;

export type ServerHandlerResult<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerResult<ServerApi, Path, Method>;

export type ServerHandlerBody<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = ServerHandlerResult<Path, Method>['body'];

export type ServerHandlerIncluded<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> =
  ServerHandlerBody<Path, Method> extends infer B
    ? B extends { readonly included?: infer I }
      ? I
      : never
    : never;

export type ServerHandlerParameter<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerParameters<ServerApi, Path, Method>;

export type ServerRequestContent<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = RequestContent<ServerApi, Path, Method>;

export function getErrorResponse<
  const T extends {
    status: keyof HttpStatusMap;
    code: string;
    detail?: string;
    source?: object;
  },
>(option: T) {
  let error: WritableDeep<Omit<T, 'status'>> & {
    status: HttpStatusMap[T['status']];
  } = { ...structuredClone(option), status: httpStatuses[option.status] };
  return {
    status: option.status as T['status'],
    body: { errors: [error] },
    contentType: 'application/json' as const,
  };
}

export type SubServerDescription<K extends string> = Pick<
  ServerDescription<ServerApi>,
  Extract<keyof ServerDescription<ServerApi>, `${K}${string}`>
>;

type GetRunResourcesOptions =
  | { filter: Parameters<Store['getRuns']>[0] }
  | {
      runs: Omit<
        Awaited<ReturnType<Store['getRuns']>>[number],
        'runCreatedAt'
      >[];
    };
export async function getRunResources(
  store: Store,
  options: GetRunResourcesOptions,
) {
  const runs =
    'runs' in options ? options.runs : await store.getRuns(options.filter);
  const [experiments, lastLogs] = await Promise.all([
    store.getExperiments({
      experimentId: pipe(
        runs,
        uniqueBy((run) => run.experimentId),
        map((run) => run.experimentId),
      ),
    }),
    store.getLastLogs({ runId: runs.map((run) => run.runId) }),
  ]);
  const groupedLastLogs = groupBy(lastLogs, (log) => log.runId);

  return {
    runs: runs.map((run) => {
      const runLastLogs = groupedLastLogs[run.runId] ?? [];
      return {
        id: run.runId,
        type: 'runs' as const,
        attributes: {
          status: run.runStatus,
          name: run.runName,
          lastLogNumber: Math.max(0, ...runLastLogs.map((l) => l.number)),
        },
        relationships: {
          lastLogs: {
            data: runLastLogs.map((log) => ({
              id: log.logId,
              type: 'logs' as const,
            })),
          },
          experiment: {
            data: { id: run.experimentId, type: 'experiments' as const },
          },
        },
      };
    }),
    experiments: experiments.map((experiment) => ({
      id: experiment.experimentId,
      type: 'experiments' as const,
      attributes: { name: experiment.experimentName },
    })),
    lastLogs: lastLogs.map((log) => ({
      id: log.logId,
      type: 'logs' as const,
      attributes: { number: log.number, logType: log.type, values: log.values },
      relationships: {
        run: { data: { id: log.runId, type: 'runs' as const } },
      },
    })),
  };
}

export function getAllowedAndFilteredRunIds(
  sessionData: SessionData['data'] | undefined,
  queryFilter: undefined | string | string[],
) {
  if (sessionData == null) {
    return [];
  }
  if (sessionData.role === 'host') {
    return queryFilter;
  }
  if (queryFilter == null) {
    return sessionData.runs;
  }
  return intersection(sessionData.runs, arrayify(queryFilter, true));
}
