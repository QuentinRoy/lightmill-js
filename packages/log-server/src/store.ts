import fs from 'node:fs/promises';
import path from 'node:path';
import * as url from 'node:url';
import SQliteDB from 'better-sqlite3';
import cuid from 'cuid';
import {
  Kysely,
  FileMigrationProvider,
  Generated,
  Migrator,
  SqliteDialect,
  CamelCasePlugin,
  DeduplicateJoinsPlugin,
} from 'kysely';
import { JsonObject } from 'type-fest';
import { arrayify } from './utils.js';
import loglevel, { LogLevelDesc } from 'loglevel';

const __dirname = url.fileURLToPath(new URL('.', import.meta.url));
const migrationFolder = path.join(__dirname, 'db-migrations');

type RunTable = {
  id: string;
  experimentId: string | null;
  createdAt: string;
  endedAt: string | null;
};
type LogTable = {
  id: Generated<bigint>;
  type: string;
  runId: string;
  createdAt: string;
};
type LogValueTable = {
  logId: bigint;
  name: string;
  value: string;
};
type SessionTable = {
  id: string;
  createdAt: string;
  expiresAt: string | null;
  runId?: string;
  role: 'admin' | 'participant';
  cookie: string;
};

type Database = {
  run: RunTable;
  log: LogTable;
  logValue: LogValueTable;
  session: SessionTable;
};

export class Store {
  #db: Kysely<Database>;
  constructor(
    db: string,
    { logLevel = loglevel.getLevel() }: { logLevel?: LogLevelDesc } = {}
  ) {
    const log = loglevel.getLogger('store');
    log.setLevel(logLevel);
    this.#db = new Kysely({
      dialect: new SqliteDialect({ database: new SQliteDB(db) }),
      log: (event) => {
        if (event.level === 'query') {
          log.debug(event.query.sql, event.query.parameters);
        } else if (event.level === 'error') {
          log.error(event.error);
        }
      },
      plugins: [new CamelCasePlugin(), new DeduplicateJoinsPlugin()],
    });
  }

  async addRun({ id, experimentId }: { id?: string; experimentId?: string }) {
    let runId = id ?? cuid();
    await this.#db
      .insertInto('run')
      .values({
        id: runId,
        createdAt: new Date().toISOString(),
        experimentId,
        endedAt: null,
      })
      .execute();
    return runId;
  }

  async endRun(id: string) {
    await this.#db
      .updateTable('run')
      .set({ endedAt: new Date().toISOString() })
      .where('id', '=', id)
      .execute();
  }

  // This methods is O(n^2) in the number of logs to insert, but n is expected
  // to be relatively small.
  async addLogs(
    logs: Array<{ type: string; runId: string; values: JsonObject }>
  ) {
    await this.#db.transaction().execute(async (trx) => {
      let createdAt = new Date().toISOString();
      let dbLogs: Array<{
        id: bigint;
        type: string;
        runId: string;
        isAssigned?: boolean;
      }> = await trx
        .insertInto('log')
        .values(
          logs.map(({ type, runId }) => {
            return { type, runId, createdAt };
          })
        )
        .returning(['id as id', 'type as type', 'runId as runId'])
        .execute();

      // Bulk insert returning values does not guarantee order, so we need to
      // match the log values logs to returned log ids.
      let dbValues = [];
      for (let log of logs) {
        let dbLog = dbLogs.find(
          (it) =>
            it.type === log.type && it.runId === log.runId && !it.isAssigned
        );
        if (!dbLog) {
          throw new Error(
            `failed to find an unassigned inserted log with type "${log.type}" and runId "${log.runId}"`
          );
        }
        dbLog.isAssigned = true;
        dbValues.push(...deconstructValues(log.values, { logId: dbLog.id }));
      }
      await trx.insertInto('logValue').values(dbValues).execute();
    });
  }

  async getLogValueNames(filter: LogFilter = {}) {
    let result = await this.#db
      .selectFrom('logValue')
      .if(filter.type != null, (qb) =>
        qb.innerJoin('log', 'log.id', 'logValue.logId').where((qb) => {
          let types = arrayify(filter.type, true);
          for (let type of types) {
            qb = qb.orWhere('log.type', '=', type);
          }
          return qb;
        })
      )
      .if(filter.runId != null, (qb) =>
        qb.innerJoin('log', 'log.id', 'logValue.logId').where((qb) => {
          let runIds = arrayify(filter.runId, true);
          for (let runId of runIds) {
            qb = qb.orWhere('log.runId', '=', runId);
          }
          return qb;
        })
      )
      .if(filter.experimentId != null, (qb) =>
        qb
          .innerJoin('log', 'log.id', 'logValue.logId')
          .innerJoin('run', 'run.id', 'log.runId')
          .where((qb) => {
            let experimentIds = arrayify(filter.experimentId, true);
            for (let experimentId of experimentIds) {
              qb = qb.orWhere('run.experimentId', '=', experimentId);
            }
            return qb;
          })
      )
      .select('logValue.name as logValueName')
      .orderBy('logValueName')
      .distinct()
      .execute();
    return result.map((it) => it.logValueName);
  }

  async hasNonEmptyExperimentId() {
    let result = await this.#db
      .selectFrom('run')
      .where('experimentId', 'is not', null)
      .select('experimentId')
      .executeTakeFirst();
    return result != null;
  }

  async *getLogs(filter: LogFilter = {}) {
    let result = await this.#db
      .selectFrom('logValue')
      .innerJoin('log', 'log.id', 'logValue.logId')
      .innerJoin('run', 'run.id', 'log.runId')
      .if(filter.type != null, (qb) =>
        qb.where((qb) => {
          let types = arrayify(filter.type, true);
          for (let type of types) {
            qb = qb.orWhere('log.type', '=', type);
          }
          return qb;
        })
      )
      .if(filter.runId != null, (qb) =>
        qb.where((qb) => {
          let runIds = arrayify(filter.runId, true);
          for (let runId of runIds) {
            qb = qb.orWhere('log.runId', '=', runId);
          }
          return qb;
        })
      )
      .if(filter.experimentId != null, (qb) =>
        qb.where((qb) => {
          let experimentIds = arrayify(filter.experimentId, true);
          for (let experimentId of experimentIds) {
            qb = qb.orWhere('run.experimentId', '=', experimentId);
          }
          return qb;
        })
      )
      .orderBy('logValue.logId')
      .select([
        'log.id as logId',
        'log.type as logType',
        'log.createdAt as logCreatedAt',
        'log.runId',
        'run.experimentId',
        'logValue.name',
        'logValue.value',
      ])
      .execute();

    let currentLog = null;
    let currentLogId = null;
    for (let row of result) {
      if (currentLog == null || row.logId !== currentLogId) {
        if (currentLog != null) {
          yield currentLog;
        }
        currentLog = {
          type: row.logType,
          experimentId: row.experimentId ?? undefined,
          runId: row.runId,
          createdAt: new Date(row.logCreatedAt),
          values: {} as JsonObject,
        };
        currentLogId = row.logId;
      }
      currentLog.values[row.name] = JSON.parse(row.value);
    }
    if (currentLog != null) {
      yield currentLog;
    }
  }

  async setSession({
    id,
    expiresAt,
    runId,
    cookie,
    role,
  }: {
    id?: string;
    expiresAt?: Date;
    runId?: string;
    role: 'admin' | 'participant';
    cookie: JsonObject;
  }) {
    // Use explicit column properties to avoid using extra properties in the
    // update set clause (e.g. createdAt could come back since it is
    // provided in getSession).
    let values = {
      id: id ?? cuid(),
      expiresAt: expiresAt?.toISOString() ?? undefined,
      runId,
      role,
      cookie: JSON.stringify(cookie),
    };
    await this.#db
      .insertInto('session')
      .values({ createdAt: new Date().toISOString(), ...values })
      .onConflict((conflict) => conflict.column('id').doUpdateSet(values))
      .execute();
  }

  async getSession(id: string) {
    let result = await this.#db
      .selectFrom('session')
      .where('session.id', '=', id)
      .leftJoin('run', 'run.id', 'session.runId')
      .select([
        'session.id as session.id',
        'session.createdAt as session.createdAt',
        'session.expiresAt as session.expiresAt',
        'session.cookie as session.cookie',
        'session.role as session.role',
        'run.id as run.id',
        'run.endedAt as run.endedAt',
        'run.createdAt as run.createdAt',
      ])
      .executeTakeFirst();
    if (!result) return;

    return {
      id: result['session.id'],
      createdAt: new Date(result['session.createdAt']),
      expiresAt: result['session.expiresAt']
        ? new Date(result['session.expiresAt'])
        : null,
      cookie: result['session.cookie']
        ? JSON.parse(result['session.cookie'])
        : null,
      role: result['session.role'],
      run:
        result['run.id'] != null
          ? {
              id: result['run.id'],
              endedAt: result['run.endedAt']
                ? new Date(result['run.endedAt'])
                : null,
              createdAt: result['run.createdAt']
                ? new Date(result['run.createdAt'])
                : null,
            }
          : null,
    };
  }

  async deleteSession(id: string) {
    await this.#db.deleteFrom('session').where('id', '=', id).execute();
  }

  async migrateDatabase() {
    let migrator = new Migrator({
      db: this.#db,
      provider: new FileMigrationProvider({
        fs,
        path,
        migrationFolder,
      }),
    });
    return migrator.migrateToLatest();
  }
}

function deconstructValues(
  data: JsonObject
): Array<{ name: string; value: string }>;
function deconstructValues<P extends Record<string, unknown>>(
  data: JsonObject,
  patch: P
): Array<P & { name: string; value: string }>;
function deconstructValues(
  data: JsonObject,
  patch?: Record<string, unknown>
): Array<JsonObject & { name: string; value: string }> {
  return Object.entries(data).map(([name, value]) => ({
    ...patch,
    name,
    value: JSON.stringify(value),
  }));
}

type LogFilter = {
  type?: string | string[];
  runId?: string | string[];
  experimentId?: string | string[];
};
