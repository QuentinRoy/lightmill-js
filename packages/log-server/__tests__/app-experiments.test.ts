import { afterEach, describe, vi } from 'vitest';
import { apiMediaType } from '../src/app-utils.ts';
import { StoreError } from '../src/store.ts';
import { arrayify } from '../src/utils.js';
import {
  apiContentTypeRegExp,
  createSessionTest,
  isMockStore,
  storeTypes,
} from './test-utils.js';

afterEach(() => {
  vi.resetAllMocks();
});

const allTests = storeTypes
  .flatMap((storeType) => [
    { storeType, sessionType: 'host' as const },
    { storeType, sessionType: 'participant' as const },
  ])
  .map((testOptions) => ({
    ...testOptions,
    test: createSessionTest(testOptions),
  }));
const hostTests = allTests.filter((op) => op.sessionType === 'host');
const participantTests = allTests.filter(
  (op) => op.sessionType === 'participant',
);

describe.for(hostTests)(
  'LogServer: post /experiments ($sessionType session, $storeType store)',
  ({ test: it }) => {
    it('creates an experiment', async ({ session: { api } }) => {
      await api
        .post('/experiments')
        .set('Content-Type', apiMediaType)
        .send({
          data: { type: 'experiments', attributes: { name: 'exp-name' } },
        })
        .expect(201, { data: { type: 'experiments', id: '1' } })
        .expect('location', 'http://lightmill-test.com/experiments/1')
        .expect('Content-Type', apiContentTypeRegExp);
    });

    it('refuses to create an experiment if there are name conflicts', async ({
      session: { api, store },
    }) => {
      if (isMockStore(store)) {
        store.addExperiment.mockImplementation(async () => {
          throw new StoreError('message', StoreError.EXPERIMENT_EXISTS);
        });
      } else {
        await store.addExperiment({ experimentName: 'exp-name' });
      }

      await api
        .post('/experiments')
        .set('Content-Type', apiMediaType)
        .send({
          data: { type: 'experiments', attributes: { name: 'exp-name' } },
        })
        .expect(409, {
          errors: [
            {
              status: 'Conflict',
              code: 'EXPERIMENT_EXISTS',
              detail: 'An experiment named "exp-name" already exists',
            },
          ],
        })
        .expect('Content-Type', apiContentTypeRegExp);
    });
  },
);

describe.for(participantTests)(
  'LogServer: post /experiments ($sessionType session, $storeType store)',
  ({ test: it }) => {
    it('refuses to create an experiment', async ({ session: { api } }) => {
      await api
        .post('/experiments')
        .set('Content-Type', apiMediaType)
        .send({
          data: { type: 'experiments', attributes: { name: 'exp-name' } },
        })
        .expect(403, {
          errors: [
            {
              status: 'Forbidden',
              code: 'FORBIDDEN',
              detail: 'Only hosts can create experiments',
            },
          ],
        })
        .expect('Content-Type', apiContentTypeRegExp);
    });
  },
);

describe.for(allTests)(
  'LogServer: get /experiments ($sessionType session, $storeType store)',
  ({ sessionType, storeType }) => {
    const it = createSessionTest({
      sessionType,
      storeType,
      setup: {
        mock: async ({ store }) => {
          const experiments = ['exp-1', 'exp-2', 'exp-3'].map((id) => ({
            experimentId: id,
            experimentName: `${id}-name`,
            experimentCreatedAt: new Date('2023-01-01T00:00:00Z'),
          }));
          const runs = experiments.map(({ experimentId }, runNum) => ({
            runId: `run-${runNum + 1}`,
            experimentId: experimentId,
            runName: `run-${runNum + 1}-name`,
            runCreatedAt: new Date('2023-01-01T00:00:00Z'),
            runStatus: 'running' as const,
          }));
          store.getExperiments.mockImplementation(async () => experiments);
          store.getRuns.mockImplementation(async (filter) => {
            if (filter?.runId != null) {
              let runFilter = arrayify(filter.runId);
              return runFilter.map((runId) => {
                let experimentId = /^(exp-\d+)/.exec(runId)?.[1] ?? 'exp-?';
                return {
                  runId,
                  experimentId,
                  runName: `${runId}-name`,
                  runCreatedAt: new Date('2023-01-01T00:00:00Z'),
                  runStatus: 'running',
                };
              });
            }
            return runs;
          });
          return { experiments, runs };
        },
        sqlite: async ({ store }) => {
          vi.useFakeTimers({ now: new Date('2023-01-01T00:00:00Z') });
          const experiments = await Promise.all(
            ['exp-1-name', 'exp-2-name', 'exp-3-name'].map((name) =>
              store.addExperiment({ experimentName: name }),
            ),
          );
          const runs = await Promise.all(
            experiments.map((exp, index) => {
              return store.addRun({
                experimentId: exp.experimentId,
                runName: `run-${index + 1}-name`,
                runStatus: 'running',
              });
            }),
          );
          vi.useRealTimers();
          return { experiments, runs };
        },
      },
    });

    it('returns the list of experiments', async ({
      session: { api, experiments },
    }) => {
      await api
        .get('/experiments')
        .expect(200, {
          data: experiments.map((exp) => ({
            id: exp.experimentId,
            type: 'experiments',
            attributes: { name: exp.experimentName },
          })),
        })
        .expect('Content-Type', apiContentTypeRegExp);
    });

    it('filters experiments by name when given filter[name] parameter', async ({
      expect,
      session: { api, store, experiments },
    }) => {
      const expNumber = 1;
      const expName = experiments[expNumber].experimentName;
      const expId = experiments[expNumber].experimentId;
      if (isMockStore(store)) {
        store.getExperiments.mockImplementation(async () => {
          return [
            {
              experimentId: expId,
              experimentName: expName,
              experimentCreatedAt: new Date('2023-01-01T00:00:00Z'),
            },
          ];
        });
      }

      await api
        .get(`/experiments?filter[name]=${expName}`)
        .expect(200, {
          data: [
            { id: expId, type: 'experiments', attributes: { name: expName } },
          ],
        })
        .expect('Content-Type', apiContentTypeRegExp);

      if (isMockStore(store)) {
        expect(store.getExperiments.mock.calls).toMatchSnapshot();
      }
    });

    it('returns empty array when no experiments match the name filter', async ({
      session: { api, store },
    }) => {
      if (isMockStore(store)) {
        store.getExperiments.mockImplementation(async (filter) => {
          if (filter?.experimentName === 'non-existent') {
            return [];
          }
          return [];
        });
      }
      await api
        .get('/experiments?filter[name]=non-existent')
        .expect(200, { data: [] })
        .expect('Content-Type', apiContentTypeRegExp);
    });
  },
);

describe.for(allTests)(
  'LogServer: get /experiments/{id} ($sessionType session, $storeType store)',
  ({ test: it }) => {
    it('returns an experiment by ID', async ({ session: { api, store } }) => {
      let experimentId = 'exp-1';
      if (isMockStore(store)) {
        store.getExperiments.mockImplementation(async (filter) => {
          if (filter?.experimentId === 'exp-1') {
            return [
              {
                experimentId,
                experimentName: 'exp-1-name',
                experimentCreatedAt: new Date('2023-01-01T00:00:00Z'),
              },
            ];
          }
          return [];
        });
      } else {
        ({ experimentId } = await store.addExperiment({
          experimentName: 'exp-1-name',
        }));
      }

      await api
        .get(`/experiments/${experimentId}`)
        .expect(200, {
          data: {
            id: experimentId,
            type: 'experiments',
            attributes: { name: 'exp-1-name' },
          },
        })
        .expect('Content-Type', apiContentTypeRegExp);
    });

    it('returns 404 for non-existent experiment', async ({
      session: { api, store },
    }) => {
      if (isMockStore(store)) {
        store.getExperiments.mockImplementation(async () => []);
      }
      await api
        .get('/experiments/non-existent')
        .expect(404, {
          errors: [
            {
              status: 'Not Found',
              code: 'EXPERIMENT_NOT_FOUND',
              detail: 'Experiment non-existent not found',
            },
          ],
        })
        .expect('Content-Type', apiContentTypeRegExp);
    });
  },
);
