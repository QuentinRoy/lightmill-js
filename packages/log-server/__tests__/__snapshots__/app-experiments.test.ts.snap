// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogServer: get /experiments ('host' session, 'mock' store) > filters experiments by name when given filter[name] parameter 1`] = `
[
  [
    {
      "experimentName": "exp-2-name",
    },
  ],
]
`;

exports[`LogServer: get /experiments ('participant' session, 'mock' store) > filters experiments by name when given filter[name] parameter 1`] = `
[
  [
    {
      "experimentName": "exp-2-name",
    },
  ],
]
`;
